src/main/java/org/assertj/core/api/AbstractObjectAssert.java

```diff
@@ -712,14 +712,7 @@ public SELF hasFieldOrPropertyWithValue(String name, Object value) {
   @CheckReturnValue
   public <ASSERT extends AbstractAssert<?, ?>> ASSERT extracting(String propertyOrField,
                                                                  InstanceOfAssertFactory<?, ASSERT> assertFactory) {
-    return internalExtracting(propertyOrField).asInstanceOf(assertFactory);
-  }
-
-  private AbstractObjectAssert<?, ?> internalExtracting(String propertyOrField) {
-    Object value = byName(propertyOrField).apply(actual);
-    String extractedPropertyOrFieldDescription = extractedDescriptionOf(propertyOrField);
-    String description = mostRelevantDescription(info.description(), extractedPropertyOrFieldDescription);
-    return newObjectAssert(value).withAssertionState(myself).as(description);
+    return extracting(propertyOrField, this::newObjectAssert).asInstanceOf(assertFactory);
```

Comment:
> That is, why create an ObjectAssert only to map it into a different assert type? It is possible the generic signature you used for `extracting` caused this impl. The generics change I suggest, would allow this fix.

> After all, the main purpose of this is to make the protected implementation public for `AbstractObjectAssert`, so it should simply delegate to the protected implementation.

The current `extracting` in `AbstractObjectAssert` is not only mapping to another assertion type, but also ensures type assertion before applying the factory, therefore the `InstanceOfAssertFactory` parameter. The type assertion was delegated to `asInstanceOf` at the expense of an intermediate (and yes, useless) `ObjectAssert`.

I tried another approach: performing the type assertion similarly to `asInstanceOf`, right before calling the protected `extracting`. I'm not convinced about it because for completeness `InstanceOfAssertFactory#getType()` should be public and not package-private anymore (not needed in AbstractObjectAssert but could be needed by third party assertion classes using the same pattern).

Honestly I prefer to keep the `extracting().asInstanceOf()` chain.